File: core/runtime/runtime-runner.js
class RuntimeRunner {
  constructor(registry = {}) {
    this.registry = registry;
  }

  async run(workflow, context = {}) {
    const state = {
      workflowId: workflow.id,
      status: 'running',
      steps: []
    };

    for (const step of workflow.steps || []) {
      const module = this.registry[step.module];
      if (!module) {
        state.status = 'failed';
        state.steps.push({ ...step, status: 'failed', error: 'module_not_found' });
        return state;
      }

      const result = await module.execute(step.action, context);
      state.steps.push({ ...step, status: 'completed', output: result });
      context = { ...context, ...result };
    }

    state.status = 'completed';
    state.output = context;
    return state;
  }
}

module.exports = RuntimeRunner;

File: core/runtime/runtime_runner.py
"""Minimal workflow runtime executor.

Runtime only coordinates modules. It does not contain AI logic.
"""

from typing import Dict, Any


class RuntimeRunner:
    def __init__(self, registry):
        self.registry = registry

    def run(self, workflow: Dict[str, Any], context=None):
        context = context or {}
        results = []

        for step in workflow.get("steps", []):
            module = self.registry.get(step["module"])
            if not module:
                raise RuntimeError(f"module not found: {step['module']}")

            result = module.execute(step.get("action"), context)
            results.append({
                "module": step["module"],
                "action": step.get("action"),
                "result": result,
            })
            context.update(result if isinstance(result, dict) else {})

        return {"status": "completed", "steps": results, "context": context}
